---
title: Earth Microbiome Vuegen Demo Notebook
fig-align: center
execute:
  echo: false
  output: asis
format:
  html:
    toc: true
    toc-location: left
    toc-depth: 3
    page-layout: full
    self-contained: true
include-in-header:
    text: |
        <style type="text/css">
        .footer {
        position: relative;
        left: 0;
        width: 100%;
        text-align: center;
        margin-top: 20px;
        }
        </style>
include-after-body:
    text: |
        <footer class="footer">
            This report was generated with 
            <a href="https://github.com/Multiomics-Analytics-Group/vuegen" target="_blank">
                <img src="https://raw.githubusercontent.com/Multiomics-Analytics-Group/vuegen/main/docs/images/vuegen_logo.svg" alt="VueGen" width="65px">
            </a>
            | © 2025 <a href="https://github.com/Multiomics-Analytics-Group" target="_blank">Multiomics Network Analytics Group (MoNA)</a>
        </footer>
---

```{python}
#| label: 'Imports'
import dataframe_image as dfi
import pandas as pd
import altair as alt
import requests
from itables import show
import plotly.io as pio
```

The Earth Microbiome Project (EMP) is a systematic attempt to characterize global microbial taxonomic and functional diversity for the benefit of the planet and humankind. 
  It aimed to sample the Earth’s microbial communities at an unprecedented scale in order to advance our understanding of the organizing biogeographic principles that govern microbial community structure. 
  The EMP dataset is generated from samples that individual researchers have compiled and contributed to the EMP. 
  The result is both a reference database giving global context to DNA sequence data and a framework for incorporating data from future studies, fostering increasingly complete characterization of Earth’s microbial diversity.
  
  You can find more information about the Earth Microbiome Project at https://earthmicrobiome.org/ and in the [original article](https://www.nature.com/articles/nature24621).

![](https://raw.githubusercontent.com/ElDeveloper/cogs220/master/emp-logo.svg){fig-alt= width=650 height=400}

# Exploratory Data Analysis
This section contains the exploratory data analysis of the Earth Microbiome Project (EMP) dataset.

## Sample Exploration
### Metadata Random Subset
```{python}
#| label: 'Metadata Random Subset 10'
#| fig-cap: ""

df = pd.read_csv('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/1_Exploratory_data_analysis/1_sample_exploration/1_metadata_random_subset.csv')
show(df, classes="display nowrap compact", lengthMenu=[3, 5, 10])
```

### Animal Samples Map
![](//home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/1_Exploratory_data_analysis/1_sample_exploration/2_animal_samples_map.png){fig-alt= width=950 height=400}

### Plant Samples Map
```{python}
#| label: 'Plant Samples Map 43'
#| fig-cap: ""

with open('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/1_Exploratory_data_analysis/1_sample_exploration/3_plant_samples_map.json', 'r') as plot_file:
    plot_json = plot_file.read()

fig_plotly = pio.from_json(plot_json)
fig_plotly.update_layout(width=950, height=500)

fig_plotly.show()
```

### Saline Samples Map
```{python}
#| label: 'Saline Samples Map 44'
#| fig-cap: ""

with open('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/1_Exploratory_data_analysis/1_sample_exploration/4_saline_samples_map.json', 'r') as plot_file:
    plot_json = plot_file.read()
fig_altair = alt.Chart.from_json(plot_json).properties(width=900, height=400)
fig_altair
```

### Physicochemical properties of the EMP samples
![](https://raw.githubusercontent.com/biocore/emp/master/methods/images/figureED1_physicochemical.png){fig-alt= width=950 height=400}

>Pairwise scatter plots of available physicochemical metadat are shown for temperature, salinity, oxygen, and pH, and for phosphate, nitrate, and ammonium

# Metagenomics
## Alpha Diversity
This subsection contains the alpha diversity analysis of the EMP dataset.

### Alpha Diversity Host Associated Samples
![](//home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/1_alpha_diversity/1_alpha_diversity_host_associated_samples.png){fig-alt= width=950 height=400}

### Alpha Diversity Free Living Samples
```{python}
#| label: 'Alpha Diversity Free Living Samples 47'
#| fig-cap: ""

with open('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/1_alpha_diversity/2_alpha_diversity_free_living_samples.json', 'r') as plot_file:
    plot_json = plot_file.read()

fig_plotly = pio.from_json(plot_json)
fig_plotly.update_layout(width=950, height=500)

fig_plotly.show()
```

## Average Copy Number
### Average Copy Number Emp Ontology Level2
![](//home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/2_average_copy_number/1_average_copy_number_emp_ontology_level2.png){fig-alt= width=950 height=400}

### Average Copy Number Emp Ontology Level3
```{python}
#| label: 'Average Copy Number Emp Ontology Level3 49'
#| fig-cap: ""

with open('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/2_average_copy_number/2_average_copy_number_emp_ontology_level3.json', 'r') as plot_file:
    plot_json = plot_file.read()

fig_plotly = pio.from_json(plot_json)
fig_plotly.update_layout(width=950, height=500)

fig_plotly.show()
```

## Nestedness
### Nestedness Random Subset
```{python}
#| label: 'Nestedness Random Subset 11'
#| fig-cap: ""

df = pd.read_csv('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/3_nestedness/1_nestedness_random_subset.csv')
show(df, classes="display nowrap compact", lengthMenu=[3, 5, 10])
```

### All Samples
```{python}
#| label: 'All Samples 50'
#| fig-cap: ""

with open('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/3_nestedness/2_all_samples.json', 'r') as plot_file:
    plot_json = plot_file.read()

fig_plotly = pio.from_json(plot_json)
fig_plotly.update_layout(width=950, height=500)

fig_plotly.show()
```

### Plant Samples
```{python}
#| label: 'Plant Samples 51'
#| fig-cap: ""

with open('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/3_nestedness/3_plant_samples.json', 'r') as plot_file:
    plot_json = plot_file.read()

fig_plotly = pio.from_json(plot_json)
fig_plotly.update_layout(width=950, height=500)

fig_plotly.show()
```

### Animal Samples
![](//home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/3_nestedness/4_animal_samples.png){fig-alt= width=950 height=400}

### Non Saline Samples
![](//home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/2_Metagenomics/3_nestedness/5_non_saline_samples.png){fig-alt= width=950 height=400}

## Shanon entropy analysis
This subsection contains the Shannon entropy analysis of the EMP dataset.

### Specificity of sequences and higher taxonomic groups for environment
![](https://raw.githubusercontent.com/biocore/emp/master/methods/images/figure4_entropy.png){fig-alt= width=950 height=400}

>a) Environment distribution in all genera and 400 randomly chosen tag sequence. b) and c) Shannon entropy within each taxonomic group.

# Network Analysis
## Phyla Association Networks
### Phyla Counts Subset
```{python}
#| label: 'Phyla Counts Subset 12'
#| fig-cap: ""

df = pd.read_csv('/home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/3_Network_analysis/1_phyla_association_networks/1_phyla_counts_subset.csv')
show(df, classes="display nowrap compact", lengthMenu=[3, 5, 10])
```

### Phyla Correlation Network With 0.5 Threshold Edgelist
**Number of nodes:** 33

**Number of edges:** 42


<div style="text-align: center;">
<iframe src="../quarto_report/static/Phyla_Correlation_Network_With_0.5_Threshold_Edgelist.html" alt="Phyla Correlation Network With 0.5 Threshold Edgelist plot" width="800px" height="630px"></iframe>
</div>

### Phyla Correlation Network With 0.5 Threshold
![](//home/runner/work/vuegen/vuegen/docs/example_data/Earth_microbiome_vuegen_demo_notebook/3_Network_analysis/1_phyla_association_networks/3_phyla_correlation_network_with_0.5_threshold.png){fig-alt= width=950 height=400}
